src/Chessboard.cpp

```diff
@@ -0,0 +1,313 @@
+    board_[0][7] = new Rook(PieceColor::White, 0, 7);
+
+    for (int i = 0; i < 8; i++) {
+        board_[1][i] = new Pawn(PieceColor::White, 1, i);
+    }
+
+    // initialize black pieces
+    board_[7][0] = new Rook(PieceColor::Black, 7, 0);
+    board_[7][1] = new Knight(PieceColor::Black, 7, 1);
+    board_[7][2] = new Bishop(PieceColor::Black, 7, 2);
+    board_[7][3] = new Queen(PieceColor::Black, 7, 3);
+    board_[7][4] = new King(PieceColor::Black, 7, 4);
+    board_[7][5] = new Bishop(PieceColor::Black, 7, 5);
+    board_[7][6] = new Knight(PieceColor::Black, 7, 6);
+    board_[7][7] = new Rook(PieceColor::Black, 7, 7);
+
+    for (int i = 0; i < 8; i++) {
+        board_[6][i] = new Pawn(PieceColor::Black, 6, i);
+    }
+}
+
+
+void Chessboard::clearBoard() {
+    for (int i = 0; i < 8; i++) {
+        for (int j = 0; j < 8; j++) {
+            if (board_[i][j] != nullptr) {
+                delete board_[i][j];
+            }
+        }
+    }
+}
+
+int Chessboard::movePiece(int startRow, int startCol, int endRow, int endCol) {
+    Piece* startPiece = board_[startRow][startCol];
+    Piece* endPiece = board_[endRow][endCol];
+
+    if (startPiece == nullptr) {
+        return 11; // no piece to move
+    }
+
+    if (startPiece->getColor() != currentPlayer_) {
+        return 12; // can't move opponent's piece
+    }
+
+    if (endPiece != nullptr && endPiece->getColor() == currentPlayer_) {
+        return 13; // can't capture own piece
+    }
+
+    if (!startPiece->isValidMove(endRow, endCol)) {
+        return 21; // invalid move
+    }
+
+    Chessboard copy = *this;
+    copy.board_[endRow][endCol] = startPiece;
+    copy.board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+
+    if (copy.isCheck(currentPlayer_)) {
+        return 31; // moving will cause checkmate
+    }
+
+    board_[endRow][endCol] = startPiece;
+    board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+    if (isCheck(currentPlayer_)) {
+        return 31; // moving will cause checkmate
+    }
+
+    board_[endRow][endCol] = startPiece;
+    board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+
+    if (isCheck(currentPlayer_)) {
+        return 41; // legal move and caused check
+    }
+    else {
+        currentPlayer_ = (currentPlayer_ == PieceColor::White) ? PieceColor::Black : PieceColor::White;
+        return 42; // legal move and next turn
+    }
+}
+
+    
+bool Chessboard::isCheck(PieceColor color) {
+// find the king
+    int kingRow = -1, kingCol = -1;
+    for (int i = 0; i < 8; i++) {
+        for (int j = 0; j < 8; j++) {
+            Piece* piece = board_[i][j];
+            if (piece != nullptr && piece->getType() == PieceType::King && piece->getColor() == color) {
+                kingRow = i;
+                kingCol = j;
+                break;
+            }
+        }

```

Comment:
Maybe it's better to keep a special variable called the position of the king instead of constantly looking for him all over the board